contracts/contracts/strategies/balancer/BalancerComposablePoolStrategy.sol

```diff
@@ -0,0 +1,135 @@
+
+contract BalancerComposablePoolStrategy is BalancerMetaPoolStrategy {
+    using SafeERC20 for IERC20;
+    using StableMath for uint256;
+
+    /* @notice position of BPT token in the Balancer's pool
+     *
+     * @dev this could be a storage variable that gets set by reading Balancer's pool
+     * tokens and comparing it to platformAddress. Seems more convenient, but we rather
+     * go for gas savings of an immutable variable.
+     */
+    uint256 public immutable bptTokenPoolPosition;
+
+    constructor(
+        BaseStrategyConfig memory _stratConfig,
+        BaseBalancerConfig memory _balancerConfig,
+        address _auraRewardPoolAddress,
+        uint256 _bptTokenPoolPosition
+    )
+        BalancerMetaPoolStrategy(
+            _stratConfig,
+            _balancerConfig,
+            _auraRewardPoolAddress
+        )
+    {
+        bptTokenPoolPosition = _bptTokenPoolPosition;
+    }
+
+    /*
+     * @dev Value that represents exit encoding where for max BPT given
+     * request exactly specifies the amount of underlying assets
+     * to be returned.
+     */
+    function _btpInExactTokensOutIndex()
+        internal
+        pure
+        override
+        returns (uint8)
+    {
+        return
+            uint8(
+                IBalancerVault
+                    .ComposablePoolExitKind
+                    .BPT_IN_FOR_EXACT_TOKENS_OUT
+            );
+    }
+
+    /*
+     * @dev Value that represents exit encoding where BPT tokens are supplied for
+     * proportional exit is required when calling a withdrawAll.
+     */
+    function _exactBptInTokensOutIndex()
+        internal
+        pure
+        override
+        returns (uint8)
+    {
+        return
+            uint8(
+                IBalancerVault
+                    .ComposablePoolExitKind
+                    .EXACT_BPT_IN_FOR_ALL_TOKENS_OUT
+            );
+    }
+
+    function _assetConfigVerification(address[] calldata _assets)
+        internal
+        view
+        override
+    {
+        require(
+            // aside from BPT token all assets must be supported
+            poolAssets.length - 1 == _assets.length,
+            "Pool assets length mismatch"
+        );
+
+        require(
+            // BPT position should be correctly configured
+            poolAssets[bptTokenPoolPosition] == platformAddress,
+            "BPT token position incorrect"
+        );
+
+        for (uint256 i = 0; i < _assets.length; ++i) {
+            require(
+                _assets[i] ==
+                    _fromPoolAsset(
+                        poolAssets[i >= bptTokenPoolPosition ? i + 1 : i]
+                    ),
+                "Pool assets mismatch"
+            );
+        }
+    }
+
+    function _getUserDataEncodedAmounts(uint256[] memory _amounts)
+        internal
+        view
+        virtual
+        override
+        returns (uint256[] memory amounts)
+    {
+        // remove the position in the array that represents the BPT
```

Comment:
I am still learning about this strategy and the parent strategies. I would like to learn more about why the BPT token amount gets removed.

Reply:
Hey sure, so the [Composable Stable Pools](https://docs.balancer.fi/concepts/pools/composable-stable.html#what-does-composable-mean)  have this particularity that they have the BPT token positioned next to other pool assets. This unlocks the ability of more easily nesting / batch swapping with other Balancer pools. 

When joining/exiting a Balancer pool (add liquidity / remove liquidity in Curve's terms) [encoded userData](https://github.com/OriginProtocol/origin-dollar/pull/1839/files/a3b31971cfa89c6dc38ab535b7dfb3a021c767d6#diff-87b55ced2ebd9369aef1aeef1ea49285185a8fda30d2d19081a6a3a6b0378b72R372) needs to be passed along the request that encodes expected tokens received after the join/exit. And that function expects the amounts not to have the BPT token amount present in the array (even though it is one of the assets in the Composable Stable Pool). 

Additionally when we are calculating the [amount of BPT expected to be received from withdrawing](https://github.com/OriginProtocol/origin-dollar/pull/1839/files/a3b31971cfa89c6dc38ab535b7dfb3a021c767d6#diff-87b55ced2ebd9369aef1aeef1ea49285185a8fda30d2d19081a6a3a6b0378b72R346) given assets the BPT token needs to be omitted from the calculation. Since we are swapping the BPT token in exchange for the pool assets. 

Meta stable pools (predecessors of Composable Stable Pools) didn't have the BPT token present as one of the pool's assets and thus the exclusion of BPT from the join/exit requests wasn't necessary. 

Does that make sense?